backend/functions/ml_logic/chatbotManager.py

```diff
@@ -0,0 +1,205 @@
+import os
+import re
+import threading
+
+import pandas as pd
+from dotenv import dotenv_values, load_dotenv
+from fb_manager.firebaseManager import FirebaseManager
+from langchain_community.chat_message_histories import ChatMessageHistory
+from langchain_core.messages import AIMessage
+from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder
+from langchain_core.runnables.history import RunnableWithMessageHistory
+from langchain_openai import AzureChatOpenAI
+from loguru import logger
+
+
+def clean_scraped_text(text: str) -> str:
+    """
+    Helper function to clean scraped text
+
+    Args:
+        text (str): scraped text
+
+    Returns
+        str: cleaned text
+    """
+
+    sentence = re.sub("'", "", text)  # Remove distracting single quotes
+    sentence = re.sub(" +", " ", sentence)  # Replace extra spaces
+    sentence = re.sub(r"\n: \'\'.*", "", sentence)  # Remove specific unwanted lines
+    sentence = re.sub(r"\n!.*", "", sentence)
+    sentence = re.sub(r"^:\'\'.*", "", sentence)
+    sentence = re.sub(r"\n", " ", sentence)  # Replace non-breaking new lines with space
+    sentence = re.sub("[^A-Za-z0-9 @]+", "", sentence)
+    return sentence
+
+
+def dataframe_to_text(df: pd.DataFrame) -> str:
+    """
+    Function to convert pandas dataframe (of scheme results) to cleaned text
+
+    Args:
+        df (pd.DataFrame): schemes in the user query for pandas dataframe
+
+    Returns:
+        str: cleaned text of information for each scheme
+    """
+    text_summary = ""
+    for _, row in df.iterrows():
+        cleanScrape = row["Scraped Text"]
+        sentence = clean_scraped_text(cleanScrape)
+
+        text_summary += f"Scheme Name: {row['Scheme']}, Agency: {row['Agency']}, Description: {row['Description']}, Link: {row['Link']}, Scraped Text from website: {sentence}\n"
+    return text_summary
+
+
+# TODO: get config from firebase
+class Config:
+    """Config class for loading .env file with chatbot API information"""
+
+    def __init__(self):
+        load_dotenv()
+
+        for key, value in dotenv_values().items():
+            setattr(self, key.lower(), value)
+
+    def __getattr__(self, item):
+        attr = os.getenv(item.upper())
+        if attr:
+            setattr(self, item.lower(), attr)
+        return attr
+
+
+class Chatbot:
+    """Singleton-patterned class for managing chatbot API"""
+
+    _instance = None
+
+    llm = None
+
+    _lock = threading.Lock()
+    chat_history = {}
```

Comment:
This will not work in production, i.e. when deployed in firebase functions. 

Serverless functions, such as Firebase Functions, are inherently stateless. Each invocation of the function, i.e. API call, is independent, and there's no guarantee that the same instance of a class will be used across different invocations. This means that any sessionID state stored in this instance variable may not persist between function calls.

Please use an external database or storage service (e.g., Firestore ) to store session state. This way the state will be persistent and accessible across different function invocations.